entity-framework/core/what-is-new/ef-core-9.0/whatsnew.md

```diff
@@ -1182,6 +1182,68 @@ protected override void Up(MigrationBuilder migrationBuilder)
 }
 ```
 
+<a name="concurrent-migrations"></a>
+
+### Protection against concurrent migrations
+
+EF9 introduces a locking mechanism to protect against multiple migration executions happening simultaneously, as that could leave the database in a corrupted state. This shouldn't happen when migrations are deployed to the production environment using [recommended methods](/ef/core/managing-schemas/migrations/applying#sql-scripts), however it is more likely to occur if migrations are applied at runtime using the [`DbContext.Database.Migrate()`](/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.migrate) method. We recommend not to apply migrations at runtime outside of the development cycle, but in some scenarios it is unavoidable, e.g. [working with using .NET Aspire projects](/dotnet/aspire/database/ef-core-migrations).
```

Comment:
Applying migrations at runtime isn't unavoidable with Aspire - it's possible to have a separate project managing migrations, and to make that container start up first, applying the migrations in a single place before the app instances are started. However, this is heavier and more complicated to set up.

We really should work with one of the Aspire people to have EF+Aspire docs either here in our docs, or in the Aspire docs, or both (or at least cross-link). /cc @SamMonoRT 

```suggestion
EF9 introduces a locking mechanism to protect against multiple migration executions happening simultaneously, as that could leave the database in a corrupted state. This doesn't happen when migrations are deployed to the production environment using [recommended methods](/ef/core/managing-schemas/migrations/applying#sql-scripts), but can happen if migrations are applied at runtime using the [`DbContext.Database.Migrate()`](/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.migrate) method. We recommend applying migrations at deployment, rather than as part of application startup, but that can result in more complicated application architectures (e.g. [when using .NET Aspire projects](/dotnet/aspire/database/ef-core-migrations).
```